### PR TITLE
Recursively delete folders in database projects

### LINKED ISSUE #14441: Cannot delete a folder from SQL Database Project when there are excluded files in it
- Azure Data Studio Version: 1.27.0-insider (user setup)
Commit: 9148aa1ed552cc89b7362f8132c2dec437ce5537

**Steps to Reproduce:**

1. Create new SQL Database Project, add folder and a file under it.
image

2. Right-click the **file** and select "Exclude from project"
image

3. Right-click the **folder** and select "Delete":
image

4. Error message appears:
image

**Expected result:**
Folder should be removed from disk and the project (this is the behavior of Visual Studio in such cases). Although I prefer the experience to be consistent with VS, the point could be made that we don't want to unintentionally remove user files. In that case we should at least exclude the folder from the project and keep it on disk. Either way, I don't think error message is the good solution here.

### PR DESCRIPTION
**Pull request by zijchen**
This PR fixes #14441. 
Use the recursive flag when calling rmdir which fixes the issue when a folder contains excluded items.

### CODE REVIEW

File: extensions/sql-database-projects/src/models/project.ts [hunk 1]
```diff
@@ -347,13 +347,10 @@ export class Project {
 	public async deleteFileFolder(entry: FileProjectEntry): Promise<void> {
 		// compile a list of folder contents to delete; if entry is a file, contents will contain only itself
 		const toDeleteFiles: FileProjectEntry[] = this.files.concat(this.preDeployScripts).concat(this.postDeployScripts).concat(this.noneDeployScripts).filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.File);
-		const toDeleteFolders: FileProjectEntry[] = this.files.filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.Folder).sort(x => -x.relativePath.length);
+		const toDeleteFolders: FileProjectEntry[] = this.files.filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.Folder);
 
 		await Promise.all(toDeleteFiles.map(x => fs.unlink(x.fsUri.fsPath)));
-
-		for (const folder of toDeleteFolders) {
-			await fs.rmdir(folder.fsUri.fsPath); // TODO: replace .sort() and iteration with rmdir recursive flag once that's unbugged
```
Comment (zijchen) [inline extensions/sql-database-projects/src/models/project.ts:355]: @Benjin I tested the recursive flag and it appears to work, but I can't find the PR that fixed it/documentation where it comes from. Do you know where I can find that?

**Comment by kisantia** [comment]
can you add a test in project.test.ts for this scenario?

**Comment by zijchen** [comment]
> can you add a test in project.test.ts for this scenario?

project.test.ts or projectTree.test.ts since we're deleting from the project tree?

**Comment by kisantia** [comment]
> > can you add a test in project.test.ts for this scenario?
> 
> project.test.ts or projectTree.test.ts since we're deleting from the project tree?

actually it might make more sense to add a new test in projectController.test.ts since that's where Benjin put the other delete tests when delete was added in https://github.com/microsoft/azuredatastudio/pull/11054. ProjectTree.test.ts is more for tests for the structure of the project tree, while tests for the functionality of the different actions is in projectController.test.ts and project.test.ts

**Review by ssreerama — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts [hunk 2]
```diff
@@ -263,7 +263,7 @@ export class Project {
 	 * Adds a folder to the project, and saves the project file
 	 * @param relativeFolderPath Relative path of the folder
 	 */
-	public async addFolderItem(relativeFolderPath: string): Promise<ProjectEntry> {
+	public async addFolderItem(relativeFolderPath: string): Promise<FileProjectEntry> {
```
Comment (zijchen) [inline extensions/sql-database-projects/src/models/project.ts:266]: Changed this to gain access to folder's URI.

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts (hunk 1 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/models/project.ts:355]: Hmm, I'm not seeing it either.  https://github.com/cjihrig/node/commits/master/lib/internal/fs/rimraf.js

If it's working and tests are passing on all platforms, then I'd say it's good to check in.

**Review by Benjin — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts [hunk 3]
```diff
@@ -347,13 +347,10 @@ export class Project {
 	public async deleteFileFolder(entry: FileProjectEntry): Promise<void> {
 		// compile a list of folder contents to delete; if entry is a file, contents will contain only itself
 		const toDeleteFiles: FileProjectEntry[] = this.files.concat(this.preDeployScripts).concat(this.postDeployScripts).concat(this.noneDeployScripts).filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.File);
-		const toDeleteFolders: FileProjectEntry[] = this.files.filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.Folder).sort(x => -x.relativePath.length);
+		const toDeleteFolders: FileProjectEntry[] = this.files.filter(x => x.fsUri.fsPath.startsWith(entry.fsUri.fsPath) && x.type === EntryType.Folder);
```
Comment (SakshiS-harma) [inline extensions/sql-database-projects/src/models/project.ts:350]: Did you try the scenario where there are multiple folder levels to be deleted?
A hierarchy like this:
image

If MessageQueueActivate.sql is excluded, then Storedprocs as well dbo folder should be excluded.

**Review by SakshiS-harma — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts (hunk 3 above)
Comment (zijchen) [reply extensions/sql-database-projects/src/models/project.ts:350]: Yes, that's actually why I made the test case so complicated. I'll try to capture this scenario as I simplify the test.

**Comment by alanrenmsft** [comment]
@zijchen  please redo the merge, looks like you didn't do it correctly.

**Review by alanrenmsft — APPROVED** [review]

**Pull request merged**